### PR TITLE
Add Unicode merge tests

### PR DESCRIPTION
**Pull request by cclauss**
Related to #4225

We need to go beyond ASCII and make sure that our merge tests properly handle Unicode characters.

### CODE REVIEW

**Comment by hornc** [comment]
@cclauss I'll take a look it this properly when I have some time. I think some of the strip accents and normalisation code is to help with finding title matches, and not for display or data store purposes, so may still be appropriate even with proper Unicode support . I'll check how each of these are used.

**Comment by mekarpeles** [comment]
This LGTM, waiting on @hornc in case there's additional context to the accent stripping mentioned in https://github.com/internetarchive/openlibrary/pull/4616#issuecomment-780052548

File: openlibrary/plugins/ol_infobase.py [hunk 1]
```diff
@@ -528,12 +528,7 @@ def normalize_edition_title(self, title):
         if not isinstance(title, six.text_type):
             return ""
 
-        # http://stackoverflow.com/questions/517923/what-is-the-best-way-to-remove-accents-in-a-python-unicode-string
-        def strip_accents(s):
-            return ''.join((c for c in unicodedata.normalize('NFD', s) if unicodedata.category(c) != 'Mn'))
-
-        norm = strip_accents(title).lower()
-        norm = norm.replace(' and ', ' ')
+        norm = title.lower().replace(' and ', ' ')
```
Comment (hornc) [inline openlibrary/plugins/ol_infobase.py:531]: Figuring out what this `norm` field is used for is needed, it looks like stripping accents is appropriate here, but I can't tell where this data goes.

I don't know how this would work for non-Latin characters, (it _may_ work for CJK / Hebrew / Arabic by stripping modifiers that might be catalogued differently?). Again, understanding what this normalisation is for would help.

File: openlibrary/catalog/merge/test_merge_marc.py [hunk 2]
```diff
@@ -114,25 +129,29 @@ def test_build_marc():
...
     foo = {'publishers': ['foo']}
     bar = {'publishers': ['bar']}
+    unicode_foo = {'publishers': ['føo']}
+    unicode_bar = {'publishers': ['bär']}
     foo2 = {'publishers': ['foo']}
     both = {'publishers': ['foo', 'bar']}
     assert compare_publisher({}, {}) == ('publisher', 'either missing', 0)
     assert compare_publisher(foo, {}) == ('publisher', 'either missing', 0)
     assert compare_publisher({}, bar) == ('publisher', 'either missing', 0)
+    assert compare_publisher(foo, unicode_foo) == ('publisher', 'mismatch', -25)
```
Comment (hornc) [inline openlibrary/catalog/merge/test_merge_marc.py:153]: I don't think this is correct, I think in some way the original purpose was to match on `foo` == `føo`
to account for variations in catalog entry. 

A real example:
https://openlibrary.org/search?q=publisher%3AS%C3%BCddeutscher+Verlag&mode=everything

vs.

https://openlibrary.org/search?q=publisher%3ASuddeutscher+Verlag&mode=everything

Shows that the publisher Süddeutscher Verlag is also catalogued (at least in our data, possibly elsewhere)  as Suddeutscher Verlag.

The original code I believe was actively trying to account for this and accept the variation as a match.

**Review by hornc — CHANGES_REQUESTED** [review]
I'm still working through this review, but it's going to take some more work. There are a number of 'normalize' methods in the code, some look like exact duplicates in terms of code and intent, others look they are normalizing for different purposes.

From what I've seen so far I'm not even convinced the normalize used by the import to check existing works is using the correct one -- this was the the code path I was concerned about here since we already have concerns that the code is not correctly identifying duplicates in all cases. The code in  `add_books` looks like it has the correct intention, but the `normalize` actually used by `add_books` is the one from `merge.normalize` which seems to have a different intent ...

File: openlibrary/catalog/merge/test_merge_marc.py [hunk 3]
```diff
@@ -1,3 +1,5 @@
+# -*- coding: utf-8 -*-
```
[the remark this answers is not in the document]
Comment (cclauss) [reply openlibrary/catalog/merge/test_merge_marc.py:1]: Py2 is now gone.  That line will get auto stripped by `pyupgrade` so I will avoid doing a bunch of manual edits for this legacy Py2 clutter.  The line has no effect in Py3 because all .py files are assumed to be utf-8.

**Review by hornc — CHANGES_REQUESTED** [review]
This PR touches a number of unrelated functional areas which made it hard to review. Some of the added tests reveal there is likely a problem with the current code and the way it normalizes titles for the purposes of recognising existing matches.  I think the intent behind some of these methods was to explicitly remove some Unicode differences so A = a, ü = u for an object matching algorithm, nothing to do with display or storage of catalog data.

Here's my list of what I see changing in this PR:

1) Removing `catalog.add_book strip_accents()` and `normalize()` code and unit tests
    Unfortunately add_book doesn't make use of these methods, it uses `catalog.utils.mk_norm()`, which uses `merge.normalize.normalize(s)`
    This is potentially a problem -- the add_book code looks like it does something better than `mk_norm` -- either way the intent is not made clear to evaluate the two 
    approaches. Despite add_book's `normalize()` being unused, it looks like it has more relevant `add_book` intent than the `catalog.utils`, which seems a more generic version.
    The previous tests correctly tested for the desired behaviour of the methods, which deliberately stripped accented characters from Unicode for the existing edition/work comparison algorithm. These stripped strings weren't being saved anywhere. **<<< Despite this code not being used by add_book, it looks like it should be, given the tests for 5 don't appear to be doing the right thing.**

2) 1 test case confirming the `add_book.split_subtitle()` can accept and handle Unicode, this data IS saved, and is a different function but in the same import area as 1 **<<< GOOD!**
3) 1 test case confirming new authors with Unicode in names are added, this data IS saved, and is a different function but in the same import area as 1 **<<< GOOD!**

4) 2 changes to UTF8 strings that don't really change anything about the tests.
test_load_with_redirected_author test doesn't really test anything with Unicode, utf8 strings are comapared and found to be identical as before.
test_existing_work again is just changing identical string comparison, doesn't really add anything to the test.
 
5) Adds more tests for catalog.merge.merge_marc.build_titles()  These are only used for edition matching in openlibrary.catalog.add_book.load(), these title variants are not saved anywhere. These tests only confirm the current behaviour of catalog.merge.merge_marc.build_titles() since no changes are made to the existing code here, only tests. This looks like a version of 1 which is actually used, but possibly not as complete?  **<<< GOOD to add more tests, but suggest the existing code is probably doing the wrong things and not identifying matches when it should :(**

6) Removes an unused strip_accents() from catalog.utils **<<< GOOD!**

7) A change to plugins/ol_infobase.py affecting how a custom / computed field `normalized_title_` is generated -- this looks like it IS SAVED to Infobase and relates to SOLR indexing. I don't know how this changes things. This is a completely different thing from 1-6, and 8. @cdrini may know what to do with this. I would be cautious about making changes here, it likely needs review in the context of how SOLR handles Unicode comparisons.

8) Changes to openlibrary/plugins/upstream/tests/test_merge_authors.py  -- this is the Merge authors functionality, which is completely separate from 1-6, and 7. The changes with Unicode in the values add something, but I'm not sure that the Unicode keys and Infobase Thing ids help, it just confuses things. Author merge isn't the place to test whether Infobase accepts Unicode Thing identifiers and URLs. Better representative values for `/authors/a` `/authors/b` and `/authors/c` would be `/authors/OL100A', '/authors/OL200A'  '/authors/OL300A' which do a better job of indicating that these are supposed to be Open Library author identifiers, not arbitrary strings.

**Final:**
* 2,3, and 6 are good changes to Import tests --> Landed in ~#4871~
* 1 and 5 suggest there is an underlying problem in how the import code normalizes Unicode titles when checking for exact matches. Figuring out what needs to happen then ensuring the code and tests are cover the right behavior is needed here.
* 7 should be made separately. Figuring out how it affects SOLR and search first is important. I don't know how `normalized_title_` is used. It looks like a field indexed by SOLR.
* 8 changes could be made separately, I don't think it will break anything, but I'm not convinced the majority of the changes there add value

File: openlibrary/catalog/merge/test_merge_marc.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hornc) [reply openlibrary/catalog/merge/test_merge_marc.py:1]: I reviewed these lines because they new additions in this PR, not legacy Py2 clutter AFAICT -- unless the Github diffing is doing something odd?

File: openlibrary/catalog/merge/test_merge_marc.py (hunk 2 above)
Comment (hornc) [reply openlibrary/catalog/merge/test_merge_marc.py:153]: Ok, I see now that this is testing existing behaviour, not making a change to that behaviour. The behaviour still looks wrong to me, but it's not the test's fault :D 
I've tried to cover this in my final review comment.
